src/enumerable/enumerable_gather.jl

```diff
@@ -0,0 +1,75 @@
+struct EnumerableGather{T,S,F,I} <: Enumerable
+    source::S
+    fields::F
+    indexFields::I
+    key::Symbol
+    value::Symbol
+end
+
+struct Not{T}
+    val::T
+end
+
+function gather(source::Enumerable, args...; key::Symbol = :key, value::Symbol = :value)
+    T = eltype(source)
+    fields = fieldnames(T)
+    F = typeof(fields)
+    if args != (false,)
+        prev = ()
+        firstArg = true
+        for arg in args
+            if typeof(arg) == Symbol
+                prev = (prev..., arg)
+            else typeof(arg) == Not{Symbol}
+                if firstArg
+                    prev = (a for a in fields if a != arg.val)
+                else
+                    prev = (a for a in prev if a != arg.val)
+                end
+            end
+            firstArg = false
+        end
+    else
+        prev = fields
+    end
+
+    I = typeof(prev)
+    return EnumerableGather{T, typeof(source), F, I}(source, fields, prev, key, value)
+end
+
+function Base.iterate(iter::EnumerableGather{T, S, F, I}) where {T, S, F, I}
+    source = iter.source
+    fields = fieldnames(T)
+    elements = Array{Any}(undef, 0)
```

Comment:
If ``EnumerableGather`` has a type parameter for the elements we are generating, we can actually type this array properly, and don't have to make it a ``Any`` vector.